### ISSUE TITLE
Crash at boot with MystCraft: "Could not find a DimensionType registered"

### ISSUE THREAD

**Issue by LemADEC** [SpongePowered/SpongeForge#2704]
**I am currently running**
- SpongeForge version: 1.12.2-2768-7.1.6-RC3616 (newer version are deleting items, see #2702)
- Forge version: 1.12.2-14.23.5.2836
- Java version: 1.8
- Operating System: CentOS

- Plugins/Mods:
SkyFactory4 4.0.2
Plugins (19): Minecraft, Minecraft Coder Pack, SpongeAPI, Sponge, AntiWDL, BetterChunkLoader, Broadcast, CatClearLag, Enjin Minecraft Plugin, GriefPrevention, LuckPerms, MagiBridge, Nucleus, Prism, Server Backup, SkyClaims, Total Economy, WrapperPing, spark
Mods (194): Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, Sponge, Advancement Book, Animal Crops, Apotheosis, AppleCore, AppleSkin, Applied Energistics 2, Aroma1997Core, Astral Sorcery, B.A.S.E, BNBGamingLib, Barrels, Drums, Storage & More, Baubles, Bed Patch, BiblioCraft, BnBGamingCore, Bonsai Trees, Bookshelf, Building Gadgets, Caliper, Calm Down Zombie Guy, Carry On, Ceramics, Chargers, ChineseWorkshop, Chisel, Chisels & Bits, Click Machine, CoFH Core, CoFH World, CodeChicken Lib, ComPatchedStorage, CommonCapabilities, Compact Machines 3, CompactStorage, Construct's Armory, ContentTweaker, Cooking for Blockheads, Corail Tombstone, CraftTweaker JEI Support, CraftTweaker2, Cucumber Library, Culinary Construct, Cyclic, Cyclops Core, Dark Utilities, Deep Mob Learning, Dimension Stages, Elevator Mod, Ender Crop, Ender Utilities, EnderCore, EnderStorage, Extended Crafting, Extra Cells 2, Extra Cells Patch, Farming for Blockheads, FastFurnace, FastWorkbench, Fence Overhaul, FindMe, Flux Networks, FoamFix, FoamFixCore, Forge Microblocks, Forge Multipart CBE, Forgiving Void, Game Stages, Guide-API, Headcrumbs, Horse Power, Hunting Dimension, HydroGel, InControl, Industrial Foregoing, Inspirations, Integrated Dynamics, Integrated Tunnels, IntegratedDynamics-Compat, IntegratedTunnels-Compat, Iron Jetpacks, Item Stages, JourneyMap, Just Enough Items, KleeSlabs, Limitless Structure Blocks, MCMultiPart, MTLib, MalisisCore, MalisisDoors, Mantle, MatterOverdrive: Legacy Edition, McJtyLib, Mekanism, MekanismGenerators, Minecraft Multipart Plugin, Mob Stages, Mod Tweaker, ModPack Basic Tools, ModPack Utilities, More Buckets, More Cauldrons, Morpheus, Mystcraft, Mystical Agradditions, Mystical Agriculture, NuclearCraft, OG Dragon+, OpenComputers, OpenComputers (Core), OpenComputers Xnet Driver, OreExcavation, Parabox, Pickle Tweaks, PigUtils, Pipe Master 2000, Placebo, PortalGun, Practical Logistics 2, Prestige, Primitive Crafting, ProjectE, RFToolsPower, Ranged Pumps, Real Filing Cabinet, RecipeStages, Redstone Flux, Resource Hogs, Rustic, Shadowfacts' Forgelin, Simple Generators, Simple Inventory sorting, Simple Storage Network, Sky Grid, SkyBonsais, Slab Machines, SlimyBoyos, Smooth Font Core, Snad, SonarCore, SpongeForge, Statues mod, Stuff A Sock In It, Super Sound Muffler, Surge, SwingThroughGrass, Sync, TallGates, TelePastries, Tell Me, Tesla Core Lib, Tesla Core Lib Registries, The Lost Cities, The Twilight Forest, The Weirding Gadget, Thermal Dynamics, Thermal Foundation, Thermal Innovation, Tinkers Tool Leveling, Tinkers' Complement, Tinkers' Construct, Tiny Progressions, TogetherForever, Topography, TorchMaster, Translocators, Tree Growing Simulator 2016, Triumph, Twitchcrumbs, Universal Modifiers, Valkyrie Lib, ViesCraft, Waddles, Waila, Waila Stages, Wall-Jump!, What Are We Looking At, Wither Crumbs, World Utils, WorldEdit, XL Food Mod, XNet, YNot, Yoyos, ZenStages, bouncy_creepers, mob_grinding_utils, sky_orchards, uppers

**Issue Description**
Crash at boot when loading MystCraft dimensions. Some warnings are also reported:
```
[03:57:41] [Server thread/INFO] [Sponge]: Checking for worlds that need to be migrated...
[03:57:41] [Server thread/INFO] [Sponge]: No worlds were found in need of migration.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST10] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST3] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST6] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST4] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST8] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST5] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/ERROR] [Sponge]: World [DIM_MYST9] has no dimension id. Report this to Sponge ASAP.
[03:57:41] [Server thread/INFO] [FML]: Injecting existing registry data into this server instance
...
[03:57:44] [Server thread/ERROR] [net.minecraft.server.MinecraftServer]: Encountered an unexpected exception
java.lang.RuntimeException: derp
	at net.minecraft.world.storage.SaveHandler.redirect$onLoadWorldInfo$bdj000(SaveHandler.java:1073) ~[bfb.class:?]
	at net.minecraft.world.storage.SaveHandler.func_75757_d(SaveHandler.java:122) ~[bfb.class:?]
	at org.spongepowered.common.world.WorldManager.loadAllWorlds(WorldManager.java:709) ~[WorldManager.class:1.12.2-2768-7.1.6-RC3616]
	at net.minecraft.server.MinecraftServer.func_71247_a(MinecraftServer.java:3541) ~[MinecraftServer.class:?]
	at net.minecraft.server.dedicated.DedicatedServer.func_71197_b(DedicatedServer.java:270) ~[nz.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:486) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
Caused by: java.lang.IllegalArgumentException: Could not find a DimensionType registered for world 'DIM_MYST8' with dim id: 8
	at org.spongepowered.common.util.FunctionalUtil.lambda$invalidArgument$0(FunctionalUtil.java:32) ~[FunctionalUtil.class:1.12.2-2768-7.1.6-RC3616]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[?:1.8.0_201]
	at net.minecraft.world.storage.WorldInfo.readSpongeNbt(WorldInfo.java:1725) ~[bfd.class:?]
	at net.minecraft.world.storage.SaveHandler.loadSpongeDatData(SaveHandler.java:624) ~[bfb.class:?]
	at net.minecraft.world.storage.SaveHandler.redirect$onLoadWorldInfo$bdj000(SaveHandler.java:1071) ~[bfb.class:?]
	... 6 more
```

https://gist.github.com/LemADEC/e3f16da8a3ba50dbe0a2e9820e4daccd

**Comment by LemADEC** [comment]
This seems a consequence of level_sponge.dat incorrectly referencing to dimensionId 8 for the overworld:
![image](https://user-images.githubusercontent.com/4776032/56498479-1c42f180-6502-11e9-8f10-6b99cd99dd2b.png)

Here's the expected values in that file:
![image](https://user-images.githubusercontent.com/4776032/56498492-37adfc80-6502-11e9-9a58-64ece7b712f5.png)

Corruption might be related to a Nucleus exception seen during the last successful boot:
```
[19:55:56] [Server thread/ERROR] [Sponge]: The Scheduler tried to run the task nucleus-S-3 owned by Plugin{id=nucleus, name=Nucleus, version=1.9.3-S7.1, description=The Ultimate Essentials Plugin., url=http://nucleuspowered.org, source=$
java.lang.IllegalStateException: world
        at io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.lambda$getNew$0(WorldDataManager.java:33) ~[WorldDataManager.class:1.9.3-S7.1]
        at java.util.Optional.orElseThrow(Optional.java:290) ~[?:1.8.0_201]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.getNew(WorldDataManager.java:33) ~[WorldDataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.getNew(WorldDataManager.java:19) ~[WorldDataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.DataManager$Loader.load(DataManager.java:117) ~[DataManager$Loader.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.DataManager$Loader.load(DataManager.java:106) ~[DataManager$Loader.class:1.9.3-S7.1]
        at com.github.benmanes.caffeine.cache.BoundedLocalCache$BoundedLocalLoadingCache.lambda$new$0(BoundedLocalCache.java:3311) ~[BoundedLocalCache$BoundedLocalLoadingCache.class:1.12.2-2768-7.1.6-RC3616]
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$14(BoundedLocalCache.java:2034) ~[BoundedLocalCache.class:1.12.2-2768-7.1.6-RC3616]
        at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1853) ~[?:1.8.0_201]
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2032) ~[BoundedLocalCache.class:1.12.2-2768-7.1.6-RC3616]
        at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2015) ~[BoundedLocalCache.class:1.12.2-2768-7.1.6-RC3616]
        at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:113) ~[LocalCache.class:1.12.2-2768-7.1.6-RC3616]
        at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:67) ~[LocalLoadingCache.class:1.12.2-2768-7.1.6-RC3616]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.DataManager.get(DataManager.java:135) ~[DataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.DataManager.get(DataManager.java:130) ~[DataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.getWorld(WorldDataManager.java:40) ~[WorldDataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.getWorld(WorldDataManager.java:45) ~[WorldDataManager.class:1.9.3-S7.1]
        at io.github.nucleuspowered.nucleus.modules.world.listeners.WorldGenListener.onWorldLoad(WorldGenListener.java:37) ~[WorldGenListener.class:1.9.3-S7.1]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_201]
        at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080) ~[?:1.8.0_201]
        at io.github.nucleuspowered.nucleus.modules.world.listeners.WorldGenListener.lambda$onStart$0(WorldGenListener.java:25) ~[WorldGenListener.class:1.9.3-S7.1]
...
```
https://gist.github.com/LemADEC/1d21772e6a4ec71076b23351b9ae1418

**Comment by simon816** [comment]
`dimensionType: topography:dim_0` that is not a vanilla dimension type, which could be why the dimension ID is not as expected.

**Comment by dualspiral** [comment]
That Nucleus error is Nucleus not being able to get the world and only affects Nucleus files. Symptom of the problem, not the cause.

**Comment by LemADEC** [comment]
I see the same Nucleus exception everytime a MystCraft dimension is created.
The one I've posted (2019-04-21-3.log 19:55:56) is the only time it appeared during boot.
At time of the issue, there was 7 MystCraft ages on that server: 3 4 5 6 8 9 10.

I've built a quick timeline and noticed something is a bit weird when saving that mod's dimensions:
2019-04-20
17:15:05 Dimension 8 created

2019-04-21
17:11:02 backup done, map isn't corrupted yet
19:53:18 during shutdown, dimension ids are correct but MystCraft ages are referring to the overworld name (Skubus instead of DIM_MYST*).
=> is that part of the issue?
```
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Stopping server
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving players
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving worlds
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'Skubus'/DIM_0
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM-1'/DIM_-1
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM1'/the_end
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM2'/Storage Cell
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM7'/twilight_forest
[19:53:18] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'Skubus'/Mystcraft
...
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [Skubus] (DIM0)
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [DIM-1] (DIM-1)
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [DIM1] (DIM1)
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [DIM2] (DIM2)
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [DIM7] (DIM7)
[19:53:18] [Server thread/INFO] [Sponge]: Unloading world [Skubus] (DIM10)
...
```
19:53:26 last successful server boot
21:11:02 backup done, map is already corrupted
23:53:28 during shutdown, dimension ids are mixed up
```
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Stopping server
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving players
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving worlds
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM_MYST8'/DIM_0
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM-1'/DIM_-1
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM1'/the_end
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM2'/Storage Cell
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM_MYST8'/Mystcraft
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM7'/twilight_forest
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM_MYST8'/Mystcraft
[23:53:28] [Server thread/INFO] [net.minecraft.server.MinecraftServer]: Saving chunks for level 'DIM_MYST8'/Mystcraft
...
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM_MYST8] (DIM0)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM-1] (DIM-1)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM1] (DIM1)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM2] (DIM2)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM_MYST8] (DIM3)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM7] (DIM7)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM_MYST8] (DIM8)
[23:53:29] [Server thread/INFO] [Sponge]: Unloading world [DIM_MYST8] (DIM9)
...
```
23:53:37 first server crash at boot

**Comment by brunyman** [comment]
with nucleus removed same thing happens:

Saving chunks for level 'world'/Mystcraft

And mystcraft works fine without sponge forge added. Same issue on Sponge 7.1.5

**Comment by LemADEC** [comment]
Issue was reproduced today with spongeforge-1.12.2-2768-7.1.6-RC3683
+ forge-1.12.2-14.23.5.2836
+ Nucleus-1.9.3-S7.1-MC1.12.2
+ mystcraft-1.12.2-0.13.7.03 (as part of SkyFactory4 4.0.3).

The timeline goes as:
2019-05-01
09:35:12 server boot, see https://gist.github.com/LemADEC/5d12e1f43ef29e9652f9ec86d81ac4f1
13:11:02 backup done, map isn't corrupted yet
13:22:00 Player A joins in DIM0
13:31:24 Player A creates DIM_MYST12
13:31:24 Sponge reports exception in nucleus on LoadWorldEvent
```
13:31:24] [Server thread/WARN] [Sponge]: /* Exception: 
[13:31:24] [Server thread/WARN] [Sponge]: /* java.lang.IllegalStateException: world
[13:31:24] [Server thread/WARN] [Sponge]: /*     io.github.nucleuspowered.nucleus.dataservices.loaders.WorldDataManager.lambda$getNew$0(WorldDataManager.java:33)
```
...
13:31:25 Player A joins the dimension
...
13:31:48 Player A times out
...
13:32:22 Player A joins directly in dimension 12
...
13:32:37 Player A times out
...
13:35:15 server shutdown

13:35:23 server boot
17:11:03 backup done, level_sponge.dat is corrupted
	level.dat Data.LevelName = Skubus
	level.dat_old Data.LevelName = Skubus
	level_sponge.dat SpongeData.dimensionId = 12
	level_sponge.dat_old SpongeData.dimensionId = 12
Server is still saving map properly apparently:
17:35:26 server shutdown, FML still refers to proper world save

17:35:34 server boot
From now on, FML refers to invalid world save:
[17:38:45] [Server thread/DEBUG] [FML]: Gathering id map for writing to world save DIM_MYST12
21:11:02 backup done, level.dat is also corrupted now
	level.dat Data.LevelName = DIM_MYST12
	level.dat_old Data.LevelName = DIM_MYST12
21:35:37 server shutdown

21:35:46 server fails to boot

To summarize:
1- there's an exception in nucleus when loading the world initially, could be unrelated
2- player timed out while joining the new MystCraft dimension after creating it
3- sponge data got corrupted, next boot this propagated to Forge/vanilla data, next boot crashed

**Comment by dualspiral** [comment]
To give some context as to what's being triggered in Nucleus after the original crash:

`io.github.nucleuspowered.nucleus.modules.world.listeners.WorldGenListener.lambda$onStart$0(WorldGenListener.java:25) ~[WorldGenListener.class:1.9.3-S7.1]`

That error you get from Nucleus is indeed Nucleus trying to access a Nucleus data file. When the server starts (or a world is loaded), code is executed to extract the UUID from the world, and ask Nucleus to get its backing data store for it.

Through some questionable decisions I made when creating the original storage system, the line [`Sponge.getServer().getWorldProperties(UUID).orElseThrow(() -> new IllegalStateException("world")).getUniqueId()`](https://github.com/NucleusPowered/Nucleus/blob/f6409c6a855401a0778686de1685250f17aeadc1/src/main/java/io/github/nucleuspowered/nucleus/dataservices/loaders/WorldDataManager.java#L33) is called. I realise that I'm asking for a UUID out when I have it, but the line is to verify that the world properties exist. The UUID that it calls is from [here](https://github.com/NucleusPowered/Nucleus/blob/284e1e5948bca8e175a214f9e87a071d05de3b52/src/main/java/io/github/nucleuspowered/nucleus/modules/world/listeners/WorldGenListener.java#L38), the code it calls into extracts its `UUID`.

So my code is taking the `UUID` from the world and then trying to get the `WorldProperties` from the Sponge world manager with this `UUID`, which returns an empty optional, which my code translates into an exception.

Minimal code to execute this will therefore be:

```java
    @Listener
    public void onWorldLoad(LoadWorldEvent event) {
        Sponge.getServer().getWorldProperties(event.getTargetWorld().getUniqueId()).orElseThrow(() -> new IllegalArgumentException("world"));
    }
```

This should work fine, but is what is throwing in this situation. Something has been corrupted and it's not been registered properly in the Sponge systems.

It doesn't explain the original bug, but should explain what is going on and why Nucleus seems to be involved (but actually isn't causing it and so can be ignored).

**Comment by DrZoddiak** [comment]
I believe I'm having this issue as well. Does this need any additional information tied to it?
It looks like it was pretty much covered, but I'm willing to test versions if need be.

**Comment by Zidane** [comment]
I have made an extremely hacky branch of SpongeForge and built out a jar. I need a volunteer to test. I've stressed this many times before but let me be perfectly clear....

BACK UP YOUR WORLD DATA. THIS IS NOT A DRILL. IF MY CODE FAILS, YOUR WORLDS ARE NUKED.

Let me know how things go.

https://www.dropbox.com/s/9g8zsjrv2f8irvi/spongeforge-1.12.2-2825-7.1.6-RC0.jar?dl=0

**Comment by Zidane** [comment]
@LemADEC @DrZoddiak 

Need some testing please.

**Comment by LemADEC** [comment]
Proposed hack fails for me:
1- Created new world with custom modpack (SMP)
2- Create a few MystCraft dimensions, connect/disconnect => ok
3- Restart server => ok
4- Try to rejoin last myscraft dimension
=> NOK (clicking the book loads DIM4 but send me from dimension 0 to 0, just a few blocks away)
Also visible in console:
```
[16:32:27] [Server thread/INFO] [warpdrive]: onPlayerChangedDimension LemADEC 0 -> 0
[16:34:06] [Server thread/INFO] [warpdrive]: onPlayerChangedDimension LemADEC 0 -> 0
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]: net.minecraft.world.MinecraftException: The save folder for world ./Spacheese/DIM_MYST4 is being accessed from another location, aborting
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.world.storage.SaveHandler.func_75762_c(SaveHandler.java:96)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.world.World.func_72906_B(World.java:3484)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.world.WorldServer.func_73042_a(WorldServer.java:1032)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.world.WorldServer.func_73044_a(WorldServer.java:2630)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at org.spongepowered.common.world.WorldManager.saveWorld(WorldManager.java:555)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.server.MinecraftServer.func_71267_a(MinecraftServer.java:3894)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.server.MinecraftServer.constant$getSaveTickInterval$zjo000(MinecraftServer.java:3847)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:687)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526)
[16:34:16] [Server thread/INFO] [STDERR]: [net.minecraft.server.MinecraftServer:func_71267_a:3896]:     at java.lang.Thread.run(Thread.java:748)
```

**Comment by Zidane** [comment]
What is a NOK?

**Comment by LemADEC** [comment]
When using the descriptive book, I expect to switch to that dimension.
Practically, after the restart, it loads the dimension (according to logs), but it doesn't move me to another dimension, it just moves me a few blocks away (maybe just the spawn point of the overworld?).
After a 2nd restart, I can't move when I join, I'm not even falling when breaking blocks below me.

**Comment by Zidane** [comment]
@phit explained how the book stuff works a bit. Keep in mind that we, especially me than the others, don't play these mods so I had no idea what you meant by using the book.

Back to your issue...I'll see what I can do but I am known for being blunt and I'll keep to my reputation by stating that Mystcraft may not be compatible with us without changes on his side. The mod does some reflection to hotswap out WorldInfo objects to his own DerivedWorldInfo child during WorldEvent.Load as well as he passes a "profiling dimension type" to Forge during World Generation to trigger a dimension to load. That dimension type is pointless to plugin devs. Hell even if I get this working for you guys, you'll have a config file for a dimension type that basically is a dud because I can't even get the right WorldInfo until I send him that load event and the World object is already completely made by that point.

I am of the firm opinion that one should have a really good reason before messing with the WorldInfo of a World. All Sponge does is remove the reliance on the Overworld's WorldInfo by giving each World it's own distinctive one (this allows each World to have it's own gamemode for example). We don't actually change the engine class to a child of WorldInfo and we go out of our way to ensure that mod events like Load/Save still get the Overworld's WorldInfo during them so per-map storage does not break. To me this is considered a "core engine object" which no normal mod (Sponge is not normal and accepts the responsibility that it isn't) should EVER change.

tl;dr I'll try to do what I can on Sponge's end but there are limits to what we can do.

**Comment by LemADEC** [comment]
I wasn't sure what level of details you needed, now I know :).

For the record, the mod does profiling of world generation off the main thread during initial boot. This profiling world is still loaded during the next boot according to the logs, but, I suppose, not really used after that.
This profiling generates empty region files on disk (with or without Sponge). It's invalid dimension id -2147483648 makes it easy to detect.
This profiling world is NOT the main world and it's not intended to run in the main thread. As such, I suggest this "profiling" to be considered NOT for plugins. It should also be excluded for the multi-threading protection that triggers when changing the world outside of the main thread.
The related option is `B:client.persave=true` in `config/mystcraft/core.cfg` configuration file.
Long story short: setting that option to false may simplify the changes on sponge side of things.


I did more extensive testing with latest (1.12.2-2825-7.1.6-RC3755) and just Mystcraft on a dedicated server. The good news is "the teleportation is still broken, it's not a side effect of your 'hack'".
Here is a more detailed test sequence you can use:
1- boot the server with a new world
2- mark spawn location with a diamond block
3- move a few blocks away (just to see the difference)

4- place a lectern, open it, place a "Descriptive book" in the lectern, left click the preview on the top right of the book (it's black rectangle initially)
note: this creates the first MystCraft age named DIM_MYST2, with the small cobblestone platform at the spawn point
=> this teleports you inside DIM_MYST2, at the age spawn point

5a- in the age, explore 4 chunks around to find a "star fissure" (it's a vertical hole going down to Y=0 with a starfield rendering like TheEnd portal, see picture)
![image](https://user-images.githubusercontent.com/4776032/58838887-4ab6ff00-8660-11e9-98dd-e39b652d8d08.png)
5b- jump into a "star fissure"
=> you're back in the overworld

note: if you can't find a "star fissure", repeat from step 4 to create another age from the current one.
note: you can also use a plugin to go back to the overworld, but it won't test the "star fissure" way of teleportation.

6- open the lectern, click the preview again
=> you're back in DIM_MYST2
7- go back to the overworld

8- disconnect, reconnect
=> you're in the overworld
9- open the lectern, click the preview again
=> you're back in DIM_MYST2
10- disconnect, reconnect
=> you're in DIM_MYST2

11- go back to the overworld
=> you're in the overworld
12- restart the server, connect
=> you're in the overworld
*BUG here:* intermittently, you're no longer able to move. Changing dimension with a plugin may help to recover.
13- open the lectern, click the preview again
=> you're be back in DIM_MYST2
*BUG here:* intermittently, you're actually in the overworld on the diamond block (spawn point), rebooting doesn't recover.

14- restart the server, connect
=> you're in DIM_MYST2
15- repeat from step 11


Note: this assumes book teleportation works the same in all forms (linking book vs descriptive book, lectern vs entity vs inventory forms).
Note: void didn't kill me in survival, it might be related to the issue or just one my plugins, I didn't check further.

**Comment by Zidane** [comment]
The issue with just ignoring the profiling dim type is the fact that I have to do a separate code path to handle the world he is loading and somehow never feed it back to plugins. The WorldManager is definitely not meant to separate that distinction.

**Comment by Zidane** [comment]
As promised I gave this a legitimate shot and I don't see how this will ever be compatible with Sponge

...in 1.12. The problem is that even once I can get the dimension loading (which I did and logged into it and started the server back up), the next issue is it is loading chunks async during world gen. Even if I could just "ignore" the profiling world, this would require severe re-wiring of how Sponge handles worlds to pull it off. To be frank, this is not how this works and is not how any of this is supposed to work.

...in 1.14, wgen is async so he can get away with this. Inherently, by the base game's design, Mystcraft will be made more compatible.

I am closing this for now. This doesn't mean I have given up completely but, instead, I need to figure out if 1.12 will happen for compatibility or focus on breaking changes in 1.14 to pull it off.

**Comment by no-identd** [comment]
Sorry for the necro, but, over here:
https://github.com/XCompWiz/Mystcraft-Issues/issues/379#issuecomment-578620541

@TimTinkers points out that they managed to get Mystcraft to work with SpongeForge by 'downgrading' to SpongeForge 7.1.7-RC3775. However, that's actually newer than the version which this issue was reported against originally by @LemADEC. However… They DO use an older Forge build, namely, 2487. So, technically, that makes this a regression, no? I do however wonder how this could possibly work (which apparently it does!) given what @Zidane & @gabizou (in #3029) said.

(Cc @XCompWiz, who presumably hasn't seen this thread at all so far.)

**Comment by gabizou** [comment]
So, just to reiterate, I'd **love** to see us being officially compatible with Mystcraft, but since I haven't looked into the issue in well over a year myself, I know that @Zidane did spend a fair amount of time trying to wire the world handling bit to be compatible with Mystcraft, and because some of the key objects (WorldInfo to name one) were being delegated/checked against, it was deemed not possible to get it working as Mystcraft expects. 

At this time, I'm fairly limited with being able to test 1.12 (I have the priority of working on our 1.14 update, which is a fair amount of work comparatively), but I could be looking into the compatibility issue down the road if no one else does. That being said, it's ideally something we'll have rectified in 1.14's branch due to the nature of how world handling has changed between the MC versions themselves.

**Comment by TimTinkers** [comment]
Just chiming in from @no-identd's tag with my details should they be useful for anybody else:
- I am using SpongeForge 1.12.2-2838-7.1.7-RC3776.jar now,
- I am using Mystcraft 1.12.2-0.10.668-BETA
- I am using Forge 1.12.2-14.23.5.2847

This is the highest Sponge version combination I've found that works. It works phenomenally. The only reason I even know about this combination is because I stumbled into this setup when I decided to configure my server a long time ago. Given the particular issues of Mystcraft world loading and how it breaks 1.12.2 for Sponge, it sounds like it's a miracle that it works at all.

On a side note, I just want to say that I appreciate you all on the Sponge team doing the best you can. Realistically, I've settled into 1.12.2 for the long haul given all that has been built on my server and the mods I love which have been discontinued. I wish Mystcraft were open source so there were a bit more we in the community could do.

It might some day be worthwhile to check changes within Sponge "break" this accidental, previously-undocumented compatibility between some time after 3776.

**Comment by gabizou** [comment]
Would

> * I am using SpongeForge 1.12.2-2838-7.1.7-RC3776.jar now,

Would it be possible (I know it's asking a lot, but it'll definitely narrow down some of the changes that took place) to see which builds it breaks with? Depending on which one, we might be able to do some workarounds, but it's something I'd be able to do down the line, not anything immediately.

**Comment by TimTinkers** [comment]
@gabizou 
> Would it be possible to see which builds it breaks with?

Certainly, happy to help. I tested the following builds and came to the following conclusions:
- 3776, which my server has been running for a long time, works perfectly.
- 3777 also works perfectly.
- 3782 also works perfectly.
- 3783 does not crash, but causes the player to teleport to the overworld spawn when attempting to use a Mystcraft descriptive book instead of teleporting to the Mystcraft world.
- 3784 behaves the same way 3783 does.
- 3785 once more works perfectly. This is what I've currently gone and updated my server to. AFAIK this is now the highest version which works without issues.
- 3787 was the first version I found that crashed on startup.
- 3788 also crashed on startup. This is as far as I tested.

**Comment by TimTinkers** [comment]
Update: 3785 began exhibiting the same behavior as 3783 and 3784. I downgraded to 3784 and now everything is working perfectly there. I either got versions crossed in my initial post, or something is nondeterministic.

**Comment by XCompWiz** [comment]
Poking my head in to say I'm open to collaborating on resolving compatibility issues for future release.

**Comment by LemADEC** [comment]
JustEnoughDimensions is compatible, so you can check how they do it maybe?

**Comment by MaegnusTheBlackMage** [comment]
I would like to just note that I have successfully gotten Mystcraft 0.13.7.06 Working with Sponge 2838-7.2.0 and forge 2854 by editing `global.conf` and setting `load-on-startup` to true.

Though on adding either IC2 or Pages to the server it results in a crash as worldprovidersurface cannot be cast to mystcrafts world profiling.

**Comment by Mortivarus** [comment]
Just to add to this: in mystcraft-1.12.2-0.13.7.06, in the core.cfg, there's an option to shut off baselining calculations and instead run a config file. Just set  B:useconfigs=true, this has (for now) solved crashes during start-up.
